### PR TITLE
Update/Remove Embedded Label Bounds on Shape Moved/Resized

### PR DESCRIPTION
**Pull request by Skaiir**
Direct issue is #1537
https://github.com/camunda/camunda-modeler/issues/2566
https://github.com/camunda/camunda-modeler/issues/2591

Doing a little write-up to get some ideas and feedback from people here. Note, this is a draft and is not trying to satisfy our definition of done yet. I'm mostly looking for input on implementation here.

## The core problem:

Because we position embedded labels either centrally or center-top (as in expanded SubProcesses), we do not store their bounds information in `di`. This is not a problem for us, but causes issues with other vendors who do rely on these bounds for positioning.

Looking at the positioning of a task with such bounds for example, let move it a few grid points down:

Before:

```xml
<bpmndi:BPMNShape id="Activity_0cq51bw_di" bpmnElement="Activity_0cq51bw">
  <dc:Bounds x="160" y="130" width="100" height="80" />
    <bpmndi:BPMNLabel labelStyle="LabelStyle_0_0" color:color="#d38300">
    <dc:Bounds x="346.96" y="225.35" width="109.98" height="50.05" />
  </bpmndi:BPMNLabel>
</bpmndi:BPMNShape>
```

After:

```xml
<bpmndi:BPMNShape id="Activity_0cq51bw_di" bpmnElement="Activity_0cq51bw">
  <dc:Bounds x="160" y="410" width="100" height="80" />
    <bpmndi:BPMNLabel labelStyle="LabelStyle_0_0" color:color="#d38300">
    <dc:Bounds x="346.96" y="225.35" width="109.98" height="50.05" />
  </bpmndi:BPMNLabel>
</bpmndi:BPMNShape>
```

The internal bounds remain untouched. 

This is not a major problem on individual items, but it does become a larger one when align-to-origin is run, as any vendor using explicit bounds will see their labels displaced relative to the tasks and subProcesses. 

So optimally we would want to handle these `explicitly bounded embedded labels`, but only when they are present. As in, we should still be able to work with their implicit counterpart.

## Implementation requirements:

There are three scenarios we need to consider here.

### Displacement of an activity

This is actually a pretty simple change, we can just listen to `shape.move` inside of the `BpmnUpdater` and propagate whatever delta a shape receives to the label bounds (if it has them). This applies to tasks and sub-processes, and is implemented already in the PR. 

### Resizing of a sub-process

This is a little more complicated, as we align things differently based on the `collapsed` flag of the sub-process shape. I added a helper function to calculate the "proper" positioning here. This is attached to the `shape.resize` event in the updater as well, and is somewhat hacky feeling. I'd really be up for suggestions / improvements here. The maths might also be a bit rough. 

### Changing the label itself (help wanted here)

Now this is the major problem here. The embedded label does not have an associated shape, therefore there's no serialized information on its actual bounds accessible. The only place where we actually do calculate the width and height (which are the only things we need here) is at the rendering stage. What this means is that we have no data to work from.

To fix this, there's only two options in my mind, either recalculate the bounds with the textRenderer, which from talks with @philippfromme seems to be an expensive operation, and also duplication. Or, we somehow cache them, but propagating renderer information up to the BpmnUpdater just feels really clunky. This is the main issue here, and the reason for the draft PR. 

## Some other considerations

When morphing sub-processes to tasks, or tasks to other kinds of tasks, we loose the label info entirely. Would we want to extend handling to these cases as well? It does feel weird that morphing a task with a colored label removes the styling (you can try this with the example diagram). 

## Test diagram
[embedded bounded labels.txt](https://github.com/bpmn-io/bpmn-js/files/7942321/embedded.bounded.labels.txt)

### CODE REVIEW

**Comment by marstamm** [comment]
Have we considered removing the bounds completely from embedded labels? Those are valid according to the BPMN spec and how we handle newly created Tasks since #1540 .

**Comment by MaxTru** [comment]
What about fixing the (rather easy?) case of `shape.move` for now, to resolve the MIWG issue and track the other stuff in a seperate issue. => incremental progress

File: lib/features/modeling/BpmnUpdater.js [hunk 1]
```diff
@@ -171,6 +179,49 @@ export default function BpmnUpdater(
     }
   });
 
+  // Handle embedded labels with defined bounds, to stay compatible with other vendors
+  this.executed('shape.move', function(event) {
+    var shape = event.context.shape;
+
+    if (hasExplicitlyBoundedEmbeddedLabel(shape)) {
```
Comment (nikku) [inline lib/features/modeling/BpmnUpdater.js:186]: If we hook in prior to the DI update we#ll be able to diff bounds (shape bounds vs. bounds stored in DI).

No need for separate logic to handle shape move (do / undo) as well as resize (do / undo).

File: lib/features/modeling/BpmnUpdater.js [hunk 2]
```diff
@@ -20,6 +20,14 @@ import {
   is
 } from '../../util/ModelUtil';
 
+import {
+  hasExplicitlyBoundedInternalLabel as hasExplicitlyBoundedEmbeddedLabel
```
Comment (nikku) [inline lib/features/modeling/BpmnUpdater.js:24]: If a utility has no use outside of a single file (which seems to be the case here), leave it inside that single file.

We prevent utility creep this way.

File: lib/util/LabelUtil.js [hunk 3]
```diff
@@ -31,6 +31,18 @@ export function isLabelExternal(semantic) {
...
 }
 
+/**
+ * Returns true if the given shape contains an internal label with explicit bounds
+ *
+ * @param {BpmnShape} shape
+ * @returns {boolean} true if is bounded embedded label
+ */
+export function hasExplicitlyBoundedInternalLabel(shape) {
+  return (is(shape, 'bpmn:Task') || is(shape, 'bpmn:SubProcess'))
```
Comment (nikku) [inline lib/util/LabelUtil.js:41]: is('bpmn:Activity')

**Comment by nikku** [comment]
An interesting approach @marstamm mentioned:If we move/resize/update an element, reset the embedded label (no bounds). Other tool vendors must support it and it is the most simple solution for us. 

It does conflict with our notion to play nicely with others. I'd therefor spend a little bit of time to [consider small improvements to this PR](https://github.com/bpmn-io/bpmn-js/pull/1586#discussion_r792795822) and just make it work for others, out of the box.

**Comment by Skaiir** [comment]
Okay so after quite a bit of back and fourths. This PR does the following things:

Adjusts internal label bounds when a move operation occurs. 
Clears internal label bounds when a resize operation occurs, I have some magic code in the bank that I git stashed and will put on a personal branch shortly which does do the resize operation, but it might be too much complexity. 

Sets up tests to make sure the two above things operate properly. 

@nikku I decided to keep the move handling because the issue with align-to-origin would reset so many things on other vendors diagrams. I think this is the safest thing to do. 

Related to: https://github.com/camunda/camunda-modeler/issues/2591

Following merge I will update this as well to reflect the current state - https://github.com/bpmn-io/bpmn-js/issues/1537

File: lib/features/modeling/BpmnUpdater.js [hunk 4]
```diff
@@ -341,11 +345,34 @@ BpmnUpdater.prototype.updateParent = function(element, oldParent) {
...
 BpmnUpdater.prototype.updateBounds = function(shape) {
 
   var di = getDi(shape);
 
+  // We need to adjust internal bounds if they are defined for compatibility reasons
+  if (shapeHasExplicitlyBoundedInternalLabel(shape)) {
+
+    // clear explicit internal bounds on resize
+    if (shape.width != di.bounds.width || shape.height != di.bounds.height) {
+      di.label = this._bpmnFactory.createDiLabel();
```
Comment (barmac) [inline lib/features/modeling/BpmnUpdater.js:364]: This means that we also drop the colors, don't we? We should make sure no properties beyond bounds are dropped. Let's have a test case for that. I noticed you already provided a BPMN with labelStyle and color:color -> this is a good example to verify correct behaviour.

File: lib/features/modeling/BpmnUpdater.js [hunk 5]
```diff
@@ -341,11 +345,34 @@ BpmnUpdater.prototype.updateParent = function(element, oldParent) {
   this.updateDiParent(di, parentDi);
 };
 
+// Handle embedded labels with defined bounds, to stay compatible with other vendors
+function shapeHasExplicitlyBoundedInternalLabel(shape) {
+  return (is(shape, 'bpmn:Activity'))
+    && getDi(shape).label
+    && getDi(shape).label.bounds != undefined;
```
Comment (barmac) [inline lib/features/modeling/BpmnUpdater.js:352]: I believe it's enough to check if bounds are truthy (they should always be an object):

```suggestion
    && getDi(shape).label.bounds;
```

File: lib/features/modeling/BpmnUpdater.js [hunk 6]
```diff
@@ -341,11 +345,34 @@ BpmnUpdater.prototype.updateParent = function(element, oldParent) {
...
 
 BpmnUpdater.prototype.updateBounds = function(shape) {
 
   var di = getDi(shape);
 
+  // We need to adjust internal bounds if they are defined for compatibility reasons
+  if (shapeHasExplicitlyBoundedInternalLabel(shape)) {
+
+    // clear explicit internal bounds on resize
+    if (shape.width != di.bounds.width || shape.height != di.bounds.height) {
```
Comment (barmac) [inline lib/features/modeling/BpmnUpdater.js:363]: Let's use strict comparison as we don't want to do type conversion:

```suggestion
    if (shape.width !== di.bounds.width || shape.height !== di.bounds.height) {
```

**Review by barmac — CHANGES_REQUESTED** [review]
We need to also make sure we don't drop properties beside bounds when a shape is resized.

File: test/spec/features/modeling/ResizeShapeSpec.js [hunk 7]
```diff
@@ -12,6 +12,30 @@ import {
...
+  var diagramXML = require('./ResizeShapeSpec.embeddedLabel.bpmn');
+
+  var testModules = [coreModule, modelingModule];
+
+  beforeEach(bootstrapModeler(diagramXML, { modules: testModules }));
+
+  it('should clear only explicit embedded bounds', inject(function(elementRegistry, modeling) {
+
+    // given
+    var subProcess = elementRegistry.get('Activity_1'), subProcessDi = getDi(subProcess);
```
Comment (marstamm) [inline test/spec/features/modeling/ResizeShapeSpec.js:26]: Let's only use one variable assignment per line
```suggestion
    var subProcess = elementRegistry.get('Activity_1'), 
          subProcessDi = getDi(subProcess);
```

File: test/spec/features/modeling/MoveShapeSpec.js [hunk 8]
```diff
@@ -8,6 +8,38 @@ import coreModule from 'lib/core';
...
+
+  var testModules = [coreModule, modelingModule];
+
+  beforeEach(bootstrapModeler(diagramXML, { modules: testModules }));
+
+
+  it('should move explicit embedded bounds', inject(function(elementRegistry, modeling) {
+
+    // given
+    var task = elementRegistry.get('Activity_0'), taskDi = getDi(task);
```
Comment (marstamm) [inline test/spec/features/modeling/MoveShapeSpec.js:23]: Same here
```suggestion
    var task = elementRegistry.get('Activity_0'), 
          taskDi = getDi(task);
```

**Review by marstamm — COMMENTED** [review]
Works as expected. See my two comments below and then I think we can merge it

One thing I noticed is that we replace the label in a replace operation. Pretty sure this is out of scope of this PR
Recording 2022-02-08 at 14 36 18

**Comment by Skaiir** [comment]
Yup I was just waiting on the checks to go green :D, now tis done.

File: lib/features/modeling/BpmnUpdater.js [hunk 9]
```diff
@@ -24,6 +24,10 @@ import {
   isAny
 } from './util/ModelingUtil';
 
+import {
+  delta,
```
Comment (philippfromme) [inline lib/features/modeling/BpmnUpdater.js:28]:
```suggestion
  delta
```

🤡

File: lib/features/modeling/BpmnUpdater.js [hunk 10]
```diff
@@ -346,6 +350,23 @@ BpmnUpdater.prototype.updateBounds = function(shape) {
 
   var di = getDi(shape);
 
+  // adjust internal bounds if they are defined for compatibility reasons
+  if (hasExplicitlyBoundedInternalLabel(shape)) {
+
+    // clear explicit internal bounds on resize
+    if (shape.width !== di.bounds.width || shape.height !== di.bounds.height) {
+      delete di.label.bounds;
```
Comment (philippfromme) [inline lib/features/modeling/BpmnUpdater.js:358]: This action is __not__ revertable. Once you resize a sub-process undoing that action won't bring the DI bounds back. Are we aware of that?

**Comment by philippfromme** [comment]
I'm reviewing this now. 🕵🏻

**Review by philippfromme — APPROVED** [review]
I've added a couple of adjustments. Most importantly, I moved the tests to the appropriate files. As a general rule, there's one top-level `describe` per file.

Good job! I'm happy about the solution. 👍🏻

**Pull request merged**